### ISSUE TITLE
kernel.json conda environment

### ISSUE THREAD

**Issue by cel4** [ipython/ipython#7947]
I am trying to create a jupyter kernel for a specific conda environment.

So I created following kernel.json file:

```
{
 "argv": [ "/absolute/path/to/env/bin/python", "-m", "IPython.kernel",
          "-f", "{connection_file}"],
 "display_name": "sci33",
 "language": "python"
}
```

This is the error I get. Any idea what's going wrong here?

```
[I 17:32:32.732 NotebookApp] Kernel started: 6cc361e0-a25c-42d1-8b18-faff6e7febb8
Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Fatal Python error: Py_Initialize: unable to load the file system codec
ImportError: No module named 'encodings'
```

**Comment by takluyver** [comment]
I think that means Python is failing to find its own standard library. I'm not sure why that would happen, though.

**Comment by cel4** [comment]
For me it does not work for any python interpreter in a virtualenv. Is this reproducible or is it just me?

**Comment by takluyver** [comment]
Does just starting the Python interpreter with `/absolute/path/to/env/bin/python` in a terminal work?

**Comment by cel4** [comment]
yes, seems to work fine

**Comment by takluyver** [comment]
Hmm, that's bizarre. And does `/absolute/path/to/env/bin/python -m IPython` work to start IPython?

**Comment by cel4** [comment]
yes, works fine as well

**Comment by takluyver** [comment]
:confused: That should be more or less equivalent to what the notebook is doing to launch that kernel. I'm not sure why Python would fail to start in one situation but not the other.

**Comment by Carreau** [comment]
Could the env variable in the subprocess not be the same ?

**Comment by takluyver** [comment]
That could be. Are you starting the notebook from a terminal, or some other way?

**Comment by cel4** [comment]
Yes simply `ipython notebook` in my terminal.

**Comment by takluyver** [comment]
I've just tried with a virtualenv, and it works exactly as I'd expect. Have you tested with an actual virtualenv, or only with conda envs?

**Comment by minrk** [comment]
I use kernelspecs for conda envs as well, and haven't had an issue.

@cel4 can you start regular IPython this way?

    /path/to/env/bin/python -m IPython

**Comment by takluyver** [comment]
already checked that ;-) https://github.com/ipython/ipython/issues/7947#issuecomment-76999010

**Comment by minrk** [comment]
Sorry, not paying enough attention. I do use conda envs with kernelspecs regularly, so I'm not sure what would be different. Maybe check env variables?

**Comment by cel4** [comment]
Uff, something seems to be broken here...

I'm trying to troubleshoot the problem, but new weird errors coming up:

```
[I 08:53:26.102 NotebookApp] Using existing profile dir: '/Users/ch/.ipython/profile_default'
[I 08:53:26.106 NotebookApp] Using MathJax from CDN: https://cdn.mathjax.org/mathjax/latest/MathJax.js
[I 08:53:26.129 NotebookApp] Serving notebooks from local directory: /private/tmp
[I 08:53:26.130 NotebookApp] 0 active kernels
[I 08:53:26.130 NotebookApp] The IPython Notebook is running at: http://localhost:8888/
[I 08:53:26.130 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[I 08:53:31.152 NotebookApp] Creating new notebook in
[I 08:53:32.346 NotebookApp] Kernel started: f2cfa26d-4c6e-4a25-9f4d-52ff7e5e36cc
/Users/ch/miniconda/envs/sci34/bin/python: No module named IPython
[I 08:53:35.348 NotebookApp] KernelRestarter: restarting kernel (1/5)
/Users/ch/miniconda/envs/sci34/bin/python: No module named IPython
[I 08:53:38.358 NotebookApp] KernelRestarter: restarting kernel (2/5)
/Users/ch/miniconda/envs/sci34/bin/python: No module named IPython
[I 08:53:41.365 NotebookApp] KernelRestarter: restarting kernel (3/5)
/Users/ch/miniconda/envs/sci34/bin/python: No module named IPython
[W 08:53:42.399 NotebookApp] Timeout waiting for kernel_info reply from f2cfa26d-4c6e-4a25-9f4d-52ff7e5e36cc
[I 08:53:44.376 NotebookApp] KernelRestarter: restarting kernel (4/5)
WARNING:root:kernel f2cfa26d-4c6e-4a25-9f4d-52ff7e5e36cc restarted
/Users/ch/miniconda/envs/sci34/bin/python: No module named IPython
[W 08:53:47.386 NotebookApp] KernelRestarter: restart failed
[W 08:53:47.386 NotebookApp] Kernel f2cfa26d-4c6e-4a25-9f4d-52ff7e5e36cc died, removing from map.
ERROR:root:kernel f2cfa26d-4c6e-4a25-9f4d-52ff7e5e36cc restarted failed!
[W 08:53:47.399 NotebookApp] Kernel deleted before session
[W 08:53:47.400 NotebookApp] 410 DELETE /api/sessions/91849bea-a081-4e5c-b29a-47474b7d3acf (::1) 1.37ms referer=http://localhost:8888/notebooks/Untitled13.ipynb?kernel_name=testkernel
```

Where are those `/Users/ch/miniconda/envs/sci34/bin/python: No module named IPython`  messages raised?

`/Users/ch/miniconda/envs/sci34/bin/python -m IPython --version`  prints `3.0.0`.

**Comment by minrk** [comment]
Can you run the server with `--debug`?

**Comment by cel4** [comment]
So it seems that if I start ipython in my activated `sci33` conda environment, everything works fine. However it does not seem to be possible to start this kernel when I'm in another conda or virtualenv environment.

It seems that the path to the site-packages is somehow screwed. I guess printing `site.getsitepackages()` would give some weird results. Yet, I understand too little from the kernel logic to debug that.

Here is the debug log. I didn't find anything suspicious.
https://gist.github.com/cel4/f1925979dcadfbfe9322

**Comment by minrk** [comment]
Can you share your environment variables (minus anything sensitive)? I suspect there might be something funny going on there.

**Comment by cel4** [comment]
Here you go: https://gist.github.com/cel4/ad8a35d452814a573108

**Comment by cel4** [comment]
I had some time for some further investigation: I only see this problem on my mac. I cannot reproduce it on a ubuntu box.

How to reproduce it (with miniconda):

The kernel environment:
```
conda create -n kernel python=2.7 ipython-notebook
source activate kernel
ipython kernelspec install-self --user
```

The notebook environment:
```
conda create -n ipy python=3.4 ipython-notebook
source activate ipy
ipython notebook
```

Is someone on OSX and has the time to check if it's reproducible, or just me?

**Comment by cel4** [comment]
This is still bugging me - anyone with a mac here who can try to reproduce this?

**Comment by jrderuiter** [comment]
I have the same issue on my Mac. Using Anaconda 3.4 as default environment, trying to use a Python 2.7 from a conda environment as a kernel.

**Comment by cel4** [comment]
Okay, I nuked all my extra python kernels and installed new ones. This narrowed the problem down to this rather interesting observation:

When I start the notebook server in a `ipython-3.x` environment, everything works as expected. I can start start my self-defined `python2` and `python3` kernels.

When I start the notebook server in a `python-2.x` environment, I can start kernels in `py27` environments, but not kernels in any `py34` environment.

I get the `Could not find platform independent libraries <prefix>` error in this case.

But again, this issue does not seem to be reproducible on my ubuntu box. This already narrows the problem down to the point that it's not the environments themselves, but the fact that `python2` calls `python3`.

This is where I am stuck now.

**Comment by cel4** [comment]
This problem boils down to to this problem:
```
import subprocess
from subprocess import Popen
cmd = [
    '/usr/local/bin/python',
    '-m', 'IPython.kernel',
    '-f', '/tmp/test_kernel.json'
]
p = Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
out, err = p.communicate()
print(err.decode("utf-8"))
```

-- `/Users/ch/miniconda/envs/sci34/bin/python: No module named IPython`

**Comment by takluyver** [comment]
So you're telling it `/usr/local/bin/python` but it's trying to use `...envs/sci34/bin/python`? That's very strange.

**Comment by minrk** [comment]
What do you get from `file /usr/local/bin/python`?

**Comment by minrk** [comment]
And `ls -l /usr/local/bin/python`

**Comment by cel4** [comment]
I don't think it's a symlink issue. This here prints the expected path:

```python
import subprocess
from subprocess import Popen
cmd = [
    '/usr/local/bin/python',
    '-c', 'import sys;print(sys.path)',
]
p = Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
out, err = p.communicate()
print(out.decode("utf-8"))
```

>['', '/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/setuptools-17.0-py2.7.egg', '/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pip-7.0.1-py2.7.egg', '/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python27.zip', '/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7', '/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plat-darwin', '/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plat-mac', '/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plat-mac/lib-scriptpackages', '/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-tk', '/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-old', '/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-dynload', '/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages', '/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/gtk-2.0', '/Library/Python/2.7/site-packages']

`$ file /usr/local/bin/python` --> `/usr/local/bin/python: Mach-O 64-bit executable x86_64`
`ls -l /usr/local/bin/python` -->  4 lrwxr-xr-x 1 ch admin 34 Jun  5 22:03 /usr/local/bin/python -> ../Cellar/python/2.7.10/bin/python

**Comment by takluyver** [comment]
Then I'm completely mystified. The error message you showed above is clearly coming from a different interpreter.

**Comment by paulperry** [comment]
I have this problem on Mac OSX 10.9.5 and have a bit more information:  I get the error if I start the ipython notebook within the python 2.7 (non)environment, and I DON'T get the error if I start in the python3 environment.  I generate the error by opening the notebook and switching the kernel from within the notebook.  I have no PYTHONHOME env variable set .  This definitely seems like an environment issue.

Below is the full context.  I installed Anaconda-2.2.0-MacOSX-x86_64.pkg (having previously `mv anaconda oldconda`) . 
```
~>which python
/Users/paulperry/anaconda/bin/python
Python 2.7.10 |Anaconda 2.2.0 (x86_64)| (default, May 28 2015, 17:04:42) 
~>python
Python 2.7.10 |Anaconda 2.2.0 (x86_64)| (default, May 28 2015, 17:04:42) 
[GCC 4.2.1 (Apple Inc. build 5577)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://binstar.org
>>> quit()
~> conda create -n py3 python=3 anaconda
~> [...]
~> source activate py3
discarding /Users/paulperry/anaconda/bin from PATH
prepending /Users/paulperry/anaconda/envs/py3/bin to PATH
(py3)~>python
Python 3.3.5 |Anaconda 2.2.0 (x86_64)| (default, Sep  2 2014, 13:57:31) 
[GCC 4.2.1 (Apple Inc. build 5577)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> quit()
(py3)~> ipython kernelspec install-self --user
(py3)~> source deactivate
discarding /Users/paulperry/anaconda/envs/py3/bin from PATH
~> ipython kernelspec install-self --user
~>ls ~/.ipython/kernels/
python2	python3
~>
~>cat ~/.ipython/kernels/python2/kernel.json 
{
 "display_name": "Python 2", 
 "language": "python", 
 "argv": [
  "/Users/paulperry/anaconda/bin/python", 
  "-m", 
  "IPython.kernel", 
  "-f", 
  "{connection_file}"
 ]
}~>cat ~/.ipython/kernels/python3/kernel.json 
{
 "display_name": "Python 3", 
 "language": "python", 
 "argv": [
  "/Users/paulperry/anaconda/envs/py3/bin/python", 
  "-m", 
  "IPython.kernel", 
  "-f", 
  "{connection_file}"
 ]
}~>
}~>ipython notebook
[I 19:07:31.011 NotebookApp] Using existing profile dir: u'/Users/paulperry/.ipython/profile_default'
[I 19:07:31.014 NotebookApp] Using MathJax from CDN: https://cdn.mathjax.org/mathjax/latest/MathJax.js
[I 19:07:31.040 NotebookApp] The port 8888 is already in use, trying another random port.
[I 19:07:31.049 NotebookApp] Serving notebooks from local directory: /Users/paulperry
[I 19:07:31.049 NotebookApp] 0 active kernels 
[I 19:07:31.049 NotebookApp] The IPython Notebook is running at: http://localhost:8889/
[I 19:07:31.049 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[W 19:07:46.499 NotebookApp] Notebook Untitled.ipynb is not trusted
[I 19:07:46.639 NotebookApp] Kernel started: 33aec8f1-a04c-4cbf-adc1-060e06c1dc62
[I 19:07:51.048 NotebookApp] Kernel shutdown: 33aec8f1-a04c-4cbf-adc1-060e06c1dc62
[I 19:07:51.061 NotebookApp] Kernel started: 83562e3a-0b48-4ad1-8434-d43605575545
Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Fatal Python error: Py_Initialize: unable to load the file system codec
ImportError: No module named 'encodings'
[I 19:07:54.061 NotebookApp] KernelRestarter: restarting kernel (1/5)
Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Fatal Python error: Py_Initialize: unable to load the file system codec
ImportError: No module named 'encodings'
[I 19:07:57.067 NotebookApp] KernelRestarter: restarting kernel (2/5)
Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Fatal Python error: Py_Initialize: unable to load the file system codec
ImportError: No module named 'encodings'
[I 19:08:00.071 NotebookApp] KernelRestarter: restarting kernel (3/5)
Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Fatal Python error: Py_Initialize: unable to load the file system codec
ImportError: No module named 'encodings'
[W 19:08:01.067 NotebookApp] Timeout waiting for kernel_info reply from 83562e3a-0b48-4ad1-8434-d43605575545
[I 19:08:03.076 NotebookApp] KernelRestarter: restarting kernel (4/5)
WARNING:root:kernel 83562e3a-0b48-4ad1-8434-d43605575545 restarted
Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Fatal Python error: Py_Initialize: unable to load the file system codec
ImportError: No module named 'encodings'
[W 19:08:06.083 NotebookApp] KernelRestarter: restart failed
[W 19:08:06.083 NotebookApp] Kernel 83562e3a-0b48-4ad1-8434-d43605575545 died, removing from map.
ERROR:root:kernel 83562e3a-0b48-4ad1-8434-d43605575545 restarted failed!
[W 19:08:06.092 NotebookApp] Kernel deleted before session
[W 19:08:06.092 NotebookApp] 410 DELETE /api/sessions/61b30c52-7c4d-45b1-8b77-6c2a3ccf3eac (::1) 0.89ms referer=http://localhost:8889/notebooks/Untitled.ipynb
^C[I 19:12:30.925 NotebookApp] interrupted
Serving notebooks from local directory: /Users/paulperry
0 active kernels 
The IPython Notebook is running at: http://localhost:8889/
Shutdown this notebook server (y/[n])? y
[C 19:12:32.613 NotebookApp] Shutdown confirmed
[I 19:12:32.613 NotebookApp] Shutting down kernels
~>source activate py3
discarding /Users/paulperry/anaconda/bin from PATH
prepending /Users/paulperry/anaconda/envs/py3/bin to PATH
(py3)~>ipython notebook
[I 19:12:49.112 NotebookApp] Using existing profile dir: '/Users/paulperry/.ipython/profile_default'
[I 19:12:49.117 NotebookApp] Using MathJax from CDN: https://cdn.mathjax.org/mathjax/latest/MathJax.js
[I 19:12:49.149 NotebookApp] The port 8888 is already in use, trying another random port.
[I 19:12:49.159 NotebookApp] Serving notebooks from local directory: /Users/paulperry
[I 19:12:49.159 NotebookApp] 0 active kernels 
[I 19:12:49.160 NotebookApp] The IPython Notebook is running at: http://localhost:8889/
[I 19:12:49.160 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[W 19:12:53.769 NotebookApp] Notebook Untitled.ipynb is not trusted
[I 19:12:53.901 NotebookApp] Kernel started: bffc4f3e-acff-440c-aedb-d5f3f0380077
[I 19:12:57.874 NotebookApp] Kernel shutdown: bffc4f3e-acff-440c-aedb-d5f3f0380077
[I 19:12:57.885 NotebookApp] Kernel started: ea0cb9d5-ff3e-48fc-aaff-452ff0833e8f

```

All is good under the python3 environment !  This is really going to confuse new users of ipynb .

**Comment by takluyver** [comment]
If this is the same issue as in the earlier case, it should be fixed when we release 3.2, by #8527

**Comment by minrk** [comment]
I expect it is. Closing here.
